### PR TITLE
feat(new_metrics): implement the metric entity & its prototype

### PR DESCRIPTION
**Pull request by empiredan**
https://github.com/apache/incubator-pegasus/issues/925

### CODE REVIEW

File: src/utils/metrics.cpp [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    ref_ptr<metric_entity> entity(new metric_entity(id, std::move(attrs)));
+
+    return entity;
+}
+
+ref_ptr<metric_entity> metric_entity_prototype::instantiate(const std::string &id) const
+{
+    metric_entity::attr_map attrs;
+
+    return instantiate(id, attrs);
```
Comment (acelyc111) [inline src/utils/metrics.cpp:49]:
```suggestion

    return instantiate(id, {});
```
Can be replaced by this?

File: src/utils/metrics.cpp [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+}
+
+metric_entity::~metric_entity() {}
+
+ref_ptr<metric_entity> metric_entity_prototype::instantiate(const std::string &id,
+                                                            metric_entity::attr_map attrs) const
+{
+    dassert_f(attrs.find("entity") == attrs.end(), "{}'s attribute \"entity\" is reserved", id);
+
+    attrs["entity"] = _name;
```
Comment (acelyc111) [inline src/utils/metrics.cpp:39]: Why add an 'entity' key to indicate its name? where will it be used?

File: include/dsn/utility/metrics.h [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+// METRIC_DEFINE_entity(my_entity);
+//
+// To use the entity type, declare it at the top of any .h/.cpp file (not within any namespace):
+// METRIC_DECLARE_entity(my_entity);
+//
+// Instantiating the entity in whatever class represents the entity:
+// entity_instance = METRIC_ENTITY_my_entity.instantiate(my_entity_id, ...);
+
+// Define a new entity type.
+#define METRIC_DEFINE_entity(name) ::dsn::metric_entity_prototype METRIC_ENTITY_##name(#name)
```
Comment (Shuo-Jia) [inline include/dsn/utility/metrics.h:42]:
```suggestion
#define METRIC_DEFINE_entity(name) dsn::metric_entity_prototype METRIC_ENTITY_##name(#name)
```

File: src/utils/metrics.cpp (hunk 2 above)
Comment (empiredan) [reply src/utils/metrics.cpp:39]: > Why add an 'entity' key to indicate its name? where will it be used?

Later while we are taking snapshots from metrics which will be emitted to the monitoring system such as Prometheus,  all attributes will be set as labels. `entity` is one of them and its value is **server**, **table** or **replica**. Thus it's added here.

File: include/dsn/utility/metrics.h (hunk 3 above)
Comment (empiredan) [reply include/dsn/utility/metrics.h:42]: The prepended scope resolution operator can ensure that `::dsn::metric_entity_prototype` be resolved from global namespace even if  `METRIC_DEFINE_entity` is used within a namespace. Thus I think we can keep it.

File: src/utils/metrics.cpp (hunk 2 above)
Comment (acelyc111) [reply src/utils/metrics.cpp:39]: Why not use `_name` directly in this case?

**Review by acelyc111 — APPROVED** [review]

**Review by Shuo-Jia — APPROVED** [review]

**Pull request merged**